File: 68.TextJustification/TJ.ts
function fullJustify(words: string[], maxWidth: number): string[] {
    const result: string[] = [];
    let currentLine: string[] = [];
    let currentLength = 0;

    for (const word of words) {
        if (currentLength + word.length + currentLine.length > maxWidth) {
            for (let i = 0; i < maxWidth - currentLength; i++) {
                currentLine[i % (currentLine.length - 1 || 1)] += ' ';
            }
            result.push(currentLine.join(''));
            currentLine = [];
            currentLength = 0;
        }
        currentLine.push(word);
        currentLength += word.length;
    }

    result.push(currentLine.join(' ') + ' '.repeat(maxWidth - currentLength - currentLine.length + 1));
    return result;
}

File: 68.TextJustification/TJ.py
class Solution:
    def fullJustify(self, words: List[str], maxWidth: int) -> List[str]:
        res = []
        cur_line = []
        num_of_letters = 0
        
        for word in words:
            if num_of_letters + len(word) + len(cur_line) > maxWidth:
                for i in range(maxWidth - num_of_letters):
                    cur_line[i % (len(cur_line) - 1 or 1)] += ' '
                res.append(''.join(cur_line))
                cur_line, num_of_letters = [], 0
            
            cur_line.append(word)
            num_of_letters += len(word)
        
        res.append(' '.join(cur_line).ljust(maxWidth))
        return res